### PR TITLE
feat: address book search

### LINKED ISSUE #835: Address book search
## What is the feature about

Adding search functionality to the address book.

## The list of requirements

- Add search field to `AddressBookHeader` (in the same manner as `SafeAppsHeader`).
- Filter `addressBookEntries` based on search text.
- Include `PagePlaceholder` when filtered `addressBookEntries.length === 0` (`props.img` must react to dark mode).

## Designs/sketches

image

## Links

- [Figma design](https://www.figma.com/file/ptTs6lDBeUuLNySroJ5PiF/Web-Master-File?node-id=1%3A6)

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Resolves #835

## How this PR fixes it

A search bar has been added to the `AddressBookHeader` which filters the results shown in the `AddressBookTable`. Said search bar, as well as the `action` buttons are responsive.

## How to test it

Open the address book, ensuring entries exist. Enter a search query and observe the entries filter depending it. When none are found, observe the `PagePlaceholder`.

## Screenshots

image

image

image

image

image

image

### CODE REVIEW

File: src/pages/transactions/tx.tsx [hunk 1]
```diff
@@ -12,7 +12,7 @@ const SingleTransaction: NextPage = () => {
       </Head>
 
       <main>
-        <Typography variant="h3" fontWeight={700} mb={3}>
+        <Typography variant="h3" fontWeight={700} pt={1} mb={3}>
```
Comment (iamacook) [inline src/pages/transactions/tx.tsx:15]: The title was not level with the other headers.

File: src/components/address-book/AddressBookHeader/index.tsx [hunk 2]
```diff
@@ -11,11 +15,46 @@ import ImportIcon from '@/public/images/address-book/import.svg'
...
+  disabled,
+  children,
+}: {
+  icon: ElementType
+  onClick: () => void
+  disabled?: boolean
+  children: string
+}): ReactElement => {
+  const theme = useTheme()
+  const isMobile = useMediaQuery(theme.breakpoints.down('md'))
```
Comment (katspaugh) [inline src/components/address-book/AddressBookHeader/index.tsx:30]: Can we avoid this? E.g. by using vh/vw units or a media query in CSS?

File: src/components/address-book/AddressBookTable/index.tsx [hunk 3]
```diff
@@ -62,8 +65,12 @@ const AddressBookTable = () => {
 
   const addressBook = useAddressBook()
   const addressBookEntries = Object.entries(addressBook)
+  const filteredEntries = addressBookEntries.filter(([address, name]) => {
+    const query = searchQuery.toLowerCase()
+    return address.toLowerCase().includes(query) || name.toLowerCase().includes(query)
+  })
```
Comment (katspaugh) [inline src/components/address-book/AddressBookTable/index.tsx:71]:
```suggestion
  const filteredEntries = searchQuery ? addressBookEntries.filter(([address, name]) => {
    const query = searchQuery.toLowerCase()
    return address.toLowerCase().includes(query) || name.toLowerCase().includes(query)
  }) : addressBookEntries
```

**Review by katspaugh — COMMENTED** [review]
Cool stuff!

File: src/components/address-book/AddressBookHeader/index.tsx (hunk 2 above)
Comment (iamacook) [reply src/components/address-book/AddressBookHeader/index.tsx:30]: I'd sooner reference the breakpoints via the `theme` as it matches the `Grid` logic below.

File: src/components/address-book/AddressBookTable/index.tsx [hunk 4]
```diff
@@ -62,8 +65,12 @@ const AddressBookTable = () => {
 
   const addressBook = useAddressBook()
   const addressBookEntries = Object.entries(addressBook)
+  const filteredEntries = addressBookEntries.filter(([address, name]) => {
```
Comment (schmanu) [inline src/components/address-book/AddressBookTable/index.tsx]: I think we should we memoize the filteredEntries.

**Review by schmanu — COMMENTED** [review]
Very nice addition! 💯

File: src/components/address-book/AddressBookHeader/index.tsx (hunk 2 above)
Comment (katspaugh) [reply src/components/address-book/AddressBookHeader/index.tsx:30]: How many pixels is `md`? We use `(max-width: 600px)` elsewhere in the app for mobile.

Comment (katspaugh) [reply src/components/address-book/AddressBookHeader/index.tsx:30]: I just don't see the point doing JS detection just to make an icon smaller.

The tooltip can be shown on desktop as well, no harm in that.

Comment (katspaugh) [reply src/components/address-book/AddressBookHeader/index.tsx:30]: You can't even see the tooltip, you have to tap and hold on the button to make it appear.

Comment (katspaugh) [reply src/components/address-book/AddressBookHeader/index.tsx:30]: That toolbar has usability issues on mobile, but that's beside the point.
Please don't use JS for responsiveness is my point.

**Review by katspaugh — APPROVED** [review]
👍

**Comment by katspaugh** [comment]
I've tested it on both mobile and desktop, works great! Please merge. ✅

**Pull request merged**